pyorbital/tlefile.py

```diff
@@ -366,49 +397,68 @@ def read_tle_files(self):
-                    continue
-                fnames += [path]
+        fnames = collect_fnames(paths)
+        tles = _parse_tles_for_downloader(fnames, open)
+        logging.info("Loaded %d TLEs from local files", len(tles))
 
-        tles = []
-        for fname in fnames:
-            with open(fname, 'r') as fid:
-                data = fid.read()
-            tles += self.parse_tles(data)
+        return tles
 
-        logging.info("Loaded %d TLEs from local files", len(tles))
+    def read_xml_admin_messages(self):
+        """Read Eumetsat admin messages in XML format."""
+        paths = self.config["downloaders"]["read_xml_admin_messages"]["paths"]
+        tles = read_tles_from_mmam_xml_files(paths)
+        logging.info("Loaded %d TLEs from admin message XML files", len(tles))
 
         return tles
 
-    def parse_tles(self, raw_data):
-        """Parse all the TLEs in the given raw text data."""
-        tles = []
-        line1, line2 = None, None
-        raw_data = raw_data.split('\n')
-        for row in raw_data:
-            if row.startswith('1 '):
-                line1 = row
-            elif row.startswith('2 '):
-                line2 = row
-            else:
+
+def _parse_tles_for_downloader(item, open_func):
+    return [Tle('', tle_file=io.StringIO(tle)) for tle in
+            _get_tles_from_uris(item, open_func, platform='', only_first=False)]
+
+
+def collect_fnames(paths):
+    """Collect all filenames from *paths*."""
+    fnames = []
+    for path in paths:
+        if '*' in path:
+            fnames += glob.glob(path)
+        else:
+            if not os.path.exists(path):
+                logging.error("File %s doesn't exist.", path)
                 continue
-            if line1 is not None and line2 is not None:
-                try:
-                    tle = Tle('', line1=line1, line2=line2)
-                except ValueError:
-                    logging.warning(
-                        "Invalid data found - line1: %s, line2: %s",
-                        line1, line2)
-                else:
-                    tles.append(tle)
-                line1, line2 = None, None
-        return tles
+            fnames += [path]
+    return fnames
+
+
+def read_tles_from_mmam_xml_files(paths):
+    # Collect filenames
+    fnames = collect_fnames(paths)
+    tles = []
+    for fname in fnames:
+        data = read_tle_from_mmam_xml_file(fname).split('\n')
+        for two_lines in _grouper(2, data):
+            tl_stream = io.StringIO('\n'.join(two_lines))
+            tles.append(Tle('', tle_file=tl_stream))
+    return tles
+
+
+def read_tle_from_mmam_xml_file(fname):
+    tree = ET.parse(fname)
+    root = tree.getroot()
+    data = []
+    for nav in root.findall(".//navigation"):
+        data.append(nav.find(".//line-1").text)
+        data.append(nav.find(".//line-2").text)
+
+    return "\n".join(data)
+
+
+def _grouper(n, iterable, fillvalue=None):
```

Comment:
The function name could be improved, maybe matching the docstring more?

Reply:
Will do. The naming was taken from `itertools` examples.

Reply:
Done.